diff -Naur src/CMakeLists.txt src.patched/CMakeLists.txt
--- src/CMakeLists.txt  1969-12-31 18:00:00.000000000 -0600
+++ src.patched/CMakeLists.txt  2011-06-22 11:50:25.697422244 -0500
@@ -0,0 +1,6 @@
+cmake_minimum_required(VERSION 2.6)
+include_directories(deterministicBFS common)
+add_definitions(-DCILKP)
+file(GLOB Sources deterministicBFS/*.C)
+add_executable(bfs-pbbs ${Sources})
+install(TARGETS bfs-pbbs DESTINATION bin)
