diff -Naur src/CMakeLists.txt src.patched/CMakeLists.txt
--- src/CMakeLists.txt  1969-12-31 18:00:00.000000000 -0600
+++ src.patched/CMakeLists.txt  2011-06-22 11:50:25.697422244 -0500
@@ -0,0 +1,6 @@
+cmake_minimum_required(VERSION 2.6)
+add_executable(gen_washington washington/washington.c)
+add_executable(gen_ac ac_dense/ac.c)
+add_executable(gen_ak ak/ak.c)
+add_executable(gen_rmf genrmf/genrmf.c genrmf/genio.c genrmf/genmain.c)
+install(TARGETS gen_washington gen_ac gen_ak gen_rmf DESTINATION bin)
