diff -Naur src/CMakeLists.txt src.patched/CMakeLists.txt
--- src/CMakeLists.txt  1969-12-31 18:00:00.000000000 -0600
+++ src.patched/CMakeLists.txt  2011-06-22 11:50:25.697422244 -0500
@@ -0,0 +1,24 @@
+cmake_minimum_required(VERSION 2.6)
+add_definitions(-DAdd_)
+include_directories(sprng2.0-lite/include sprng2.0-lite/src)
+set(sprng sprng2.0-lite/src/primes_32.c
+          sprng2.0-lite/src/primes_64.c
+          sprng2.0-lite/src/fwrap_mpi.c
+          sprng2.0-lite/src/cputime.c
+          sprng2.0-lite/src/makeseed.c
+          sprng2.0-lite/src/store.c
+          sprng2.0-lite/src/simple_mpi.c
+          sprng2.0-lite/src/memory.c
+          sprng2.0-lite/src/communicate.c
+          sprng2.0-lite/src/checkid.c
+          sprng2.0-lite/src/simple_mpi.c
+          sprng2.0-lite/src/sprng/sprng.c
+          sprng2.0-lite/src/lcg/lcg.c
+          sprng2.0-lite/src/lfg/lfg.c
+          sprng2.0-lite/src/lcg64/lcg64.c
+          sprng2.0-lite/src/mlfg/mlfg.c
+          sprng2.0-lite/src/cmrg/cmrg.c)
+file(GLOB rmat R-MAT/*.c)
+add_executable(GTgraph-rmat ${rmat} ${sprng})
+target_link_libraries(GTgraph-rmat m)
+install(TARGETS GTgraph-rmat DESTINATION bin)
