diff -Naur src/CMakeLists.txt src.patched/CMakeLists.txt
--- src/CMakeLists.txt  1969-12-31 18:00:00.000000000 -0600
+++ src.patched/CMakeLists.txt  2011-06-22 11:50:25.697422244 -0500
@@ -0,0 +1,24 @@
+cmake_minimum_required(VERSION 2.6)
+add_definitions(-DAdd__ -DGALOIS_PATCH)
+include_directories(sprng2.0/include sprng2.0/src)
+set(sprng 
+  sprng2.0/checkid.c
+  sprng2.0/cputime.c
+  sprng2.0/lcg64.c
+  sprng2.0/makeseed.c
+  sprng2.0/primes_32.c
+  sprng2.0/simple_mpi.c
+  sprng2.0/cmrg.c
+  sprng2.0/lcg.c
+  sprng2.0/memory.c
+  sprng2.0/primes_64.c
+  sprng2.0/sprng.c
+  sprng2.0/communicate.c
+  sprng2.0/fwrap_mpi.c
+  sprng2.0/lfg.c
+  sprng2.0/mlfg.c
+  sprng2.0/store.c)
+file(GLOB ssca *.c)
+add_executable(SSCA2 ${ssca} ${sprng})
+target_link_libraries(SSCA2 m)
+install(TARGETS SSCA2 DESTINATION bin)
diff -Naur SSCA2v2.2/init.c SSCA2/init.c
--- src/init.c    2007-08-16 15:41:26.000000000 -0500
+++ src.patched/init.c        2012-03-02 11:34:31.086524574 -0600
@@ -35,6 +35,9 @@
 
     MaxIntWeight = (1<<SCALE);
     SubGraphPathLength = 3;
+#ifdef GALOIS_PATCH
+    K4approx = SCALE;
+#else
 #ifndef VERIFYK4    
     if (SCALE < 10)
         K4approx = SCALE;
@@ -43,4 +46,5 @@
 #else
     K4approx = SCALE;
 #endif
+#endif
 }
